#164 critical pre-release hotfix in native and memory store strategy selection

jeenbroekstra opened the pull request:
This PR addresses GitHub issue: #164 .

Briefly describe the changes proposed in this PR:

* hotfix/workaround for issue with configuration of evaluation strategy in memory and native stores.
* both stores now use the provided Factory to create an EvaluationStrategy.
* EvaluationStrategy interface extended to allow passing through of `iterationCacheSyncThreshold` param.

We may want to revisit the design later, as it all is a bit tightly coupled. However, this is a hotfix and unless somehow has strong objections I'd like this merged in so we can proceed with release.

@pulquero, @ansell , your feedback in particular is appreciated.

jeenbroekstra commented:
Hudson verify: https://hudson.eclipse.org/rdf4j/job/rdf4j-verify-pr/38/

jeenbroekstra commented:
Hudson detected a compilation failure. Fix coming up.

jeenbroekstra commented:
New hudson job running: https://hudson.eclipse.org/rdf4j/job/rdf4j-verify-pr/39/

jeenbroekstra commented:
Build succeeds.

jeenbroekstra commented:
Somewhat tricky to provide a solid regression test for this as programmatic configuration of extended evaluation strategy is difficult to access. I'll cobble something together after release based on the ad-hoc test code I wrote to debug the issue but I'm guessing we will want to provide a few convenience methods later on to make this a bit easier.

pulquero commented:
In sailsourceconnection the method is correctly implemented. It is overriden in the subclasses as the sails have the sync parameter which only made sense with strict. To support sync in the general case we probably need to add setparameter/attribute methods to the factory interface or a sub interface then sail.setsync just delegates to factory.setparameter. Unfortunately I have a flight to catch so can't do anything to the respect.

jeenbroekstra commented:
No worries, and thanks for your input. 

I'm not sure why you say the parameter only makes sense with the strict strategy - it seems just as relevant for the TupleFunctionStrategy. What it controls is the behavior of internal iterations during query evaluation (in particular when sorting or grouping, which internally cache results) - it sets a threshold at which the evaluation strategy will start syncing cached results to disk to avoid out of memory errors. 

So I think there's a definite case to make this a general parameter (I'm all for getting a better name by the way: `iterationCacheSyncThreshold` is rather a mouthful). But refactoring for this is is out of scope for this particular patch - I want to get a release out today. 

I'd like to schedule issue #576, which is about a more general design for setting Sail parameters in a generic manner, for the next minor release. We can take this discussion up as part of that issue.

jeenbroekstra commented:
What I'll try and do in the short term is remove the overloaded create method from the Factory class, and instead introduce a setter for that particular parameter. That changes existing behavior less.

jeenbroekstra commented:
Cleaned up setup a bit, and added a simple regression test. 

Hudson verify running now: https://hudson.eclipse.org/rdf4j/job/rdf4j-verify-pr/40/ . If this greenlines I will merge and go ahead with release.
